### ISSUE TITLE
Allow skipping sha checksum

### ISSUE THREAD

**Issue by Yanipo5** [prisma/prisma#12230]
## Problem
Installing Prisma in an Enterprise CI (Jenkins), I encounter the issue that downloading the sha values (checkSum!?) are forbidden by company policy. That is due to a policy were by every download has do go throgh our enterprise reposetory.

## Suggested solution
Add --skipChecksum, that would prevent downloading the sha value.

## Logs
```
> npm --ignore-scripts i
> npx prisma db pull
Error: request to https://binaries.prisma.sh/all_commits/73e60b76d394f8d37d8ebd1f8918c79029f0db86/debian-openssl-1.1.x/migration-engine.gz.sha256 failed, reason: connect ECONNREFUSED 108.156.107.28:443
```

## Alternatives
1) I suppose a [mirror](https://www.prisma.io/docs/concepts/components/prisma-engines#hosting-engines) could work.
But since we are only using Prisma for tests, I haven't got to try and mirror the binaries / checkSum

2) commit all node_modules/*prisma, 
I tried that, people wasn't happy (enterprise monorepo, so it added 330Mb to all).

## Additional context
To be clear, I know the error is for pulling the sha, but I couldn't verify if the binaries were loaded,
Is there a way to confirm it?
[Open Question](https://github.com/prisma/prisma/discussions/12195)
[Requested here as well](https://github.com/prisma/prisma/issues/5105)

**Comment by janpio** [comment]
How did you get and provide the engine files? 

If you downloaded these manually and provide them explicitly via an environment variable, we should just trust you and no try to verify this via a checksum any more.

**Comment by Yanipo5** [comment]
I think prisma downloaded these binaries, and the checksum.
Could you point to which specific file / file pattern are you referring to (so I can check).
@janpio

**Comment by janpio** [comment]
The output you posted includes the failing attempt to download the checksum file: `Error: request to https://binaries.prisma.sh/all_commits/73e60b76d394f8d37d8ebd1f8918c79029f0db86/debian-openssl-1.1.x/migration-engine.gz.sha256 failed, reason: connect ECONNREFUSED 108.156.107.28:443`. I would expect the binary download to fail the same way. Can you elaborate on your networking setup that this file download is disallowed (but potentially the engine download not)?

**Comment by Yanipo5** [comment]
@janpio 
The prisma installation process runs inside a Jenkins pipeline that is setup by the enterprise policy. Therefore, I can't give networking details (I'm also unfamilliar with these).

I suspect that you are right, and that the binaries should fail to download as well,
I guess previously, I just installed with `npm --ignore-scripts i` therefore instead of failing in the installation, it failed later while trying to pull the db schema.

Here is the error from the pipeline, when I install Prisam (without ignoring scripts).
17:40:14  Error: request to https://binaries.prisma.sh/all_commits/b371888aaf8f51357c7457d836b86d12da91658b/debian-openssl-1.1.x/introspection-engine.gz.sha256 failed, reason: connect ECONNREFUSED 108.156.107.46:443

Ok, so what is the recommendation here?
Just copy the binaries into our cloud storage, and the the mirror env?

**Comment by janpio** [comment]
If you cloud storage allows you to set up proxies, you can also create one for `binaries.prisma.sh` and then configure that. Alternatively, yes you will indeed need to find a way to self host the engine files and then point to that. (A more manual variant would be to host the Engine files locally next to your installation and use other env variables to point to these directly.)

**Comment by janpio** [comment]
Related issues: https://github.com/prisma/prisma/issues/12593 + https://github.com/prisma/prisma/issues/5105

Generally, we totally agree that this behavior is not optimal. Prisma CLI unfortunately requires the checksum file in too many cases, tries to download it and then fails if that does not work. I am now aware of three cases:

1. No local cache folder for the checksum files and no internet - failure.
2. When one of the many cusotmization env variables for engine locations (`PRISMA_QUERY_ENGINE_BINARY`, `PRISMA_QUERY_ENGINE_LIBRARY`, `PRISMA_MIGRATION_ENGINE_BINARY` etc) are used and no internet to download the file - failure.
3. `PRISMA_ENGINES_MIRROR` used but a checksum file not present there (like on Artifactory which provides only 1 out of the two files we expect) - failure.

All these cases could be worked around by adding a new environment variable that ensure that the checksum checks are either completely skipped (which would also skip the request), or the request is attempted anyway but then a failure is accepted. For cases 1 and 2, the "skip the check" would probably be better, but for case 3 it is only one of them that fails because of the two, so an informative output would probably be better there.

Do you agree with this analysis @spudly?

I think we would be super happy about a PR adding this in some way. We'll just need to "design" it together to fully make sense.

**Comment by spudly** [comment]
Sounds good to me. Here's a proposed solution that (I think) resolves all 3 of the issues you listed above:

New env var: `PRISMA_IGNORE_MISSING_CHECKSUM`

When this environment property contains a truthy value, it will still attempt to download the checksum files.

If the checksum file download fails (either due to no internet or 404 error) it will log a warning and skip the checksum validation, allowing the build to continue.

If the checksum download succeeds but the checksum doesn't match the downloaded binary, the postinstall script will fail  just like it it does today.

In my case (artifactory with missing unzipped checksum file), it will fail to download the unzipped checksum file but it will still validate that the zipped checksum file matches the sha256 file from the server.

The only item I'm not sure about is #2 above. If it were me using those environment variables, I would expect it never to even attempt to validate the checksum, since I'd be providing either a custom binary or a pre-downloaded binary.

Thoughts?

**Comment by janpio** [comment]
> The only item I'm not sure about is number 2 above. If it were me using those environment variables, I would expect it never to even attempt to validate the checksum, since I'd be providing either a custom binary or a pre-downloaded binary.

That is absolutely correct, and I have hopefully misunderstood something when I wrote that list.

To your suggestion, that sounds valid but somehow the "specificity" of the env var makes me .. cautious. But I can indeed not see anything bad with it.

`PRISMA_ENGINES_CHECKSUM_IGNORE_MISSING` would be my preferred naming by the way, as that would allow an alternative `PRISMA_ENGINES_CHECKSUM_SKIP` if we ever need that for the other use case.

**Comment by nish-nair** [comment]
I am facing this exact issue with corporate servers lacking internet connectivity and artifactory is the only place to store external binaries. Agree with the above solution which is a life saver. However, at this point what is a temporary solution? checking in node_modules and do a `npm rebuild` instead of `npm i` in jenkins build?

**Comment by spudly** [comment]
I don't think we have a workaround at the moment. Even if you download the binaries and use env vars to tell prisms where they are, it still fails because it tries to download the checksum files.

I don't think checking in node modules will work unless you're deploying to the same platform you develop on.

I'm currently working on a fix (as described above) but it will likely be a week or so before I'm able to submit it as a pull request.

**Comment by thw0rted** [comment]
This might be a dumb question, but is it totally impractical to distribute the native binaries via npm?  I know everybody is moving away from `node-gyp` but is there nothing to take its place?  It would be great if we could just depend on `@prisma/query-engine-debian-openssl-1.1.x` instead of having it download the binary at runtime.  That would also neatly handle the checksum issue.

**Comment by janpio** [comment]
Not a dumb question at all @thw0rted. Our dependency on OpenSSL makes it impossible to use the built-in os specific package support of Npm and Co, and our `binaryTargets` feature makes it more complex as well (as it allows configuring the installation of multiple packages, including an auto detected `native` one, but also only of engine that are _not_ for your current platform). But we are thinking about this a lot, and hope to make progress in that direction soon. Npm would be really neat as a distribution mechanism for our engines.

**Comment by MichaelKorn** [comment]
Can anyone estimate the benefit of comparing the checksum before and after unpacking? Do you know the specific reasons for implementing both checks?
If not I would suggest just to remove the after extraction check, IMHO this is clean than requesting a .sha256 from the server for a not existing file.

**Comment by janpio** [comment]
> Can anyone estimate the benefit of comparing the checksum before and after unpacking? Do you know the specific reasons for implementing both checks?

@MichaelKorn We had trouble in the past where users had either invalid archives that could not be unpacked, and/or invalid extracted files that could not be executed. Hence checksums for both was added.

**Comment by MichaelKorn** [comment]
@janpio Thanks for the explanation.
However, I don't believe that we have a technically good solution here:

1. Requesting sha256 for non-existent files is uncommon
2. `PRISMA_ENGINES_CHECKSUM_IGNORE_MISSING` is just a workaround:
   1. This leads to warnings becoming the standard for many users
   2. The statement is that checksum checking is a useful feature, but this bypasses it even in cases where it would be needed.

The solution that I can think of that preserves the feature and solves the issue would be to write the checksum (file) for the extracted binaries into the gz archive (together with the binaries).

**Comment by thw0rted** [comment]
OK only one more dumb question, I promise:

A given version of the `@prisma/engines-version` package will always grab the same prebuilt binaries from the server, on a given target (platform/`binaryTargets` value), right?  When you build that package, do you have every possible binary it might request on hand?  If so, why not just put all the checksums for every possible binary, at that version, in the `@prisma/engines-version` NPM package?  At 86 bytes per checksum, every set of 8 checksums per target (4x binary, 4x archive) adds less than 1kb to the uncompressed package size.  Assuming Prisma supports, what, 10-12 targets?... your package will still be smaller than a lot of the READMEs in my `node_modules` folder.

**Comment by janpio** [comment]
Not a dumb question, that is definitely an option we could consider. The complexity is that in `@prisma/engines-version` we actually do _not_ have that information present ([if you are interested](https://github.com/prisma/engines-wrapper/blob/main/.github/workflows/publish-engines.yml)), so would need to gather it first... but still doable. Just not a priority right now as we assume that adding the env var right now would at least unblock everyone until we have time to really rethink this functionality.

**Comment by Jolg42** [comment]
Since [`4.16.0`](https://github.com/prisma/prisma/releases/tag/4.16.0)

From our freshly updated [docs](https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#prisma_engines_checksum_ignore_missing)

> `PRISMA_ENGINES_CHECKSUM_IGNORE_MISSING` can be can be set to a truthy value to ignore problems around downloading & verifying the integrity (via a checksum file) of the Prisma engines. This is particularly useful when deploying to an offline system environment where the checksum file cannot be downloaded.
> 
> `PRISMA_ENGINES_CHECKSUM_IGNORE_MISSING=1`
> 
> Note: we might change the overall download behavior in a future release in a way that this environment variable will not be needed anymore in a offline environment case.
